### PR TITLE
Accelerated networking nil

### PR DESCRIPTION
**Pull request by whites11**
This PR aims at enabling the AcceleratedNetworking feature when the CAPZ flag is set to nil, only when the instance type supports it.

### CODE REVIEW

**Comment by whites11** [comment]
This is still untested.

File: service/controller/internal/vmsku/vmsku.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+func (v *Interface) initCache(ctx context.Context) error {
+	v.logger.LogCtx(ctx, "level", "debug", "message", "Initializing cache for VMSKU")
+	cl, err := v.clientFactory.GetResourceSkusClient("giantswarm", "credential-default")
```
Comment (whites11) [inline service/controller/internal/vmsku/vmsku.go:77]: Is there a way to get the location of this secret in a better way?

Comment (nprokopic) [reply service/controller/internal/vmsku/vmsku.go:77]: I believe we attempt to get in from the org namespace, and then fallback to the default one, e.g. https://github.com/giantswarm/azure-operator/blob/master/client/organization_factory.go#L178-L180

File: service/controller/internal/vmsku/vmsku.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	CapabilityAcceleratedNetworking = "AcceleratedNetworkingEnabled"
+)
+
+type Config struct {
+	ClientFactory *client.Factory
+	Location      string
+	Logger        micrologger.Logger
+}
+
+type Interface struct {
```
Comment (tuommaki) [inline service/controller/internal/vmsku/vmsku.go:28]: Huh? Struct called `Interface`? Should you maybe define the interface in `spec.go` and then have implementation here for that? Then constructor below would be `func New(config Config) (Interface, error) {`

File: service/controller/internal/vmsku/vmsku.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	}, nil
+}
+
+func (v *Interface) HasCapability(ctx context.Context, vmType string, name string) (bool, error) {
+	if len(v.skus) == 0 {
+		err := v.initCache(ctx)
+		if err != nil {
+			return false, microerror.Mask(err)
+		}
+	}
```
Comment (tuommaki) [inline service/controller/internal/vmsku/vmsku.go:58]: Would it make sense to wrap this in a function like `v.ensureInititialized()` which would then do `v.initCache()` call once inside an instance specific mutex to make it read nicer and prevent concurrent init calls?

File: service/controller/internal/vmsku/vmsku.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+				if capability.Value != nil && strings.EqualFold(*capability.Value, string(CapabilitySupported)) {
+					return true, nil
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+func (v *Interface) initCache(ctx context.Context) error {
```
Comment (tuommaki) [inline service/controller/internal/vmsku/vmsku.go:75]: I think this should be somehow guarded with a mutex to only be initialized once. Alternatively one could also utilize [`sync.Once`](https://golang.org/pkg/sync/#Once).

File: service/controller/internal/vmsku/vmsku.go (hunk 1 above)
Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go:77]: here I don't really care about the organization. The outcome of this module is shared between all organizations (VM sku details do not change within the same region).
So I am really ok using the default credentials here.

File: service/controller/internal/vmsku/vmsku.go (hunk 2 above)
Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go:28]: Let me just rename it.

File: service/controller/internal/vmsku/vmsku.go (hunk 1 above)
Comment (fiunchinho) [reply service/controller/internal/vmsku/vmsku.go:76]: So there are two ways of using credentials in the operator. You either use the control plane credentials, or the tenant cluster credentials.

## Control plane creds

These are easy. They are built at the very beginning and just passed around. If you want to use these, just wire the object when is built.

## Tenant cluster creds

Without the CR we can't know which credentials to use (BYOC). So we can't just build these credentials and pass them around. They need to be built when reacting to a watched object. To build them, we need the CR to infer the credential secret.


So basically, if you don't care about which credentials are used for this particular call, just use the control plane ones. If you want to make the call with the tenant cluster credentials, your interface needs to either receive the credentials object, or the credential secret, or the CR.

File: service/controller/internal/vmsku/vmsku.go (hunk 3 above)
Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go:58]: done

File: service/controller/internal/vmsku/vmsku.go (hunk 4 above)
Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go]: done with a mutex

File: service/controller/internal/vmsku/vmsku.go (hunk 1 above)
Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go:89]: sounds good, can you link some place where we use this strategy?

Comment (fiunchinho) [reply service/controller/internal/vmsku/vmsku.go:89]: Not sure which one are you asking for. If you don't care about credentials and using the control plane is fine, maybe you can use this `AzureClientSet`? https://github.com/giantswarm/azure-operator/blob/5504970f61319f91a1e392295e85a53f79156331/service/service.go#L440

Pass it to your object in creation time and that's it.

Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go:89]: Yeah I was referring to the first strategy. Thanks

File: service/controller/internal/vmsku/vmsku.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	Location      string
+	Logger        micrologger.Logger
+}
+
+type VMSKUs struct {
+	clientFactory *client.Factory
+	location      string
+	skus          map[string]*compute.ResourceSku
+	logger        micrologger.Logger
+	mux           sync.Mutex
```
Comment (tuommaki) [inline service/controller/internal/vmsku/vmsku.go:34]: Minor nitpick: _mux_ as term is often used in context of HTTP handlers path routing e.g. [stdlib](https://golang.org/pkg/net/http/#ServeMux). Perhaps `initMutex` ?

**Review by tuommaki — APPROVED** [review]
I'm good with this now once you settle with Jose. :+1:

File: service/controller/internal/vmsku/vmsku.go (hunk 1 above)
Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go]: should be done

File: service/controller/internal/vmsku/vmsku.go (hunk 5 above)
Comment (whites11) [reply service/controller/internal/vmsku/vmsku.go]: makes a lot of sense. Done, thanks

**Review by fiunchinho — APPROVED** [review]
